Feature/friendship v1

bini-i opened the pull request:
## The following features are added:
- Friendship table with invitee_id and inviter_id columns.
- established self-join of User model to build friendship between two users.
- User model updated to include has_many associations to itself through the Friendship model.
- testing for validations, association, and instance methods in both User and Friendship models.

bini-i committed:
Add friendship migration and associations

- friendship migration with status attribute and self join association to users table

Commit:
Update readme, gitignore, ERD files
- Update Entity Relationship Diagram (ERD)
- Update readme file
- Update gitignore file

Commit:
Add Friendships table and methods
- Add methods to User model
- Create Friendships table with user id and friend id

bini-i committed:
- Add rspec gem and user_spec

- test for friends method in User model

Commit:
Rebuild friendships table and Add methods in User model
- Create friendships table with user_id and friend_id as each foreign_key
- Connect friendships with user and friend
- Add methods in User model in order to check whether they are friends or not

Commit:
Update ERD

bini-i committed:
- Add friendhsip migration

- create friendship table with invitee and inviter referencing to single
user table

bini-i committed:
- Add who_invited_me, who_i_invited through tables and friends method

- who_invited_me association goes through invitation_i_got table
- who_i_invited association goes through invitations_i_sent table
- friends method looks for all friends of a given user

Commit:
Add methods in User model
- Add approve_requests(user)
- Add friends
- Add requests_for_friendship(user)
- Add pending_requests_i_sent
- Add pending_requests_i_got
- Add friend?(user)

Commit:
Update ERD

bini-i committed:
- Add testing for User instance methods

- test added for mehthods:
	- requests_for_friendship
	- approve_request

Commit:
Add tests for User model
- Add friends, #friend?, #pending_requests_i_sent, #pending_requests_i_got

bini-i committed:
- Add validation and association tests for user in user_spec and for
friendship in friendship_spec

- added testing for validation:
	- validate length of name
- added testing for associations:
	- posts, comments, likes, invitations_i_got

bertil291utn commented on app/models/user.rb:
```diff
@@ -9,4 +9,43 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+
+  has_many :invitations_i_got, foreign_key: :invitee_id, class_name: 'Friendship'
+  has_many :who_invited_me, through: :invitations_i_got, source: :inviter
+
+  has_many :invitations_i_sent, foreign_key: :inviter_id, class_name: 'Friendship'
+  has_many :who_i_invited, through: :invitations_i_sent, source: :invitee
+
```
## One-to-many relation
- Once you have updated the User--> friendship relation with a migration, com to there and update this relation.
So far this relation is many-to-many.
_Use this [hint ](https://smartfunnycool.com/friendships-in-activerecord/)_

bertil291utn reviewed (changes requested):
# **Status:** Changes requested 🟠 
<p align="center"> <img src="https://i.cubeupload.com/4HbSFK.jpg" alt="image" width="60" /> <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master Git Flow you have a professional README file. 

However, in order to accomplish with all requirements , you required to do these changes, listed below:
### Specific requirements
- Update the User--> Friendship relation is one-to-many not many-to-many. A user can have much friendship, please read again this [tutorial 😁](https://smartfunnycool.com/friendships-in-activerecord/ ) just in **Create Has Many Relationships in the User Model** section
![image](https://user-images.githubusercontent.com/24902525/104041299-d4b53b00-51a6-11eb-9799-bba916103ed0.png)
- User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  - The user can't send an invitation, there's no button or link 
![image](https://user-images.githubusercontent.com/24902525/104040710-feba2d80-51a5-11eb-995b-79429e517ce8.png)
- These next requirements also are missing, please complete them:
  * [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  * [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  * [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  * [x] There is a button to invite user to friendship on users list page and on single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  * [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

### OPTIONAL
- It'd be great to display the current user name in the navbar section, to know which user is currently logged in 

### Summary

- In this milestone, you must have finished almost all the project requirements, check the **Project-specific** section https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/
- Also check the relation between User and friendship, in this way you're going to avoid any issues later with the next milestone 
- Don't forget to check inline comments 
***

Guys work on the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

Commit:
Update ERD files

Commit:
Add rspec test for models and Install fontawesome gem

Commit:
Update readme file

bini-i committed:
- Add friendships controller and update users show view

- friendships controller create method creats record in friendship table
for a user inviting another user
- users show view now shows a button to invite user

Commit:
Add friendship status to all users view
- Add custom style for button and status message
- Add link partial to show it in list and user show pages

Commit:
Add pending friendship invitations to user's show
- Add custom style for pending friendship invitations
- Add friendship list

Commit:
Add condition to user's show in order to display friend requests only for current_user

bini-i committed:
- update users status partial and friendships controller

- frienships contorller update action made to update accept or reject
status for friendship
- status partial made to display accept or reject status for an invite

Commit:
Add rejected status to user's view
- updated the user model with rejected method

Commit:
Fix linter errors

PhillipUg reviewed (changes requested):
Hey team, 👋

You have done an amazing job implementing the requirements for this project.

However, there are still some changes 👇 you need to implement to approve this request 🔍

### Changes Required ♻️ 

- [x] You have a failing test. Fix this
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). User shouldn't see posts from other users who are not their friend.

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

Commit:
Fix rspec test and Update timeline_posts method
- To display the posts written by current_user and friends, update timeline_posts method

descholar-ceo commented:
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

descholar-ceo commented on app/models/friendship.rb:
```diff
@@ -0,0 +1,12 @@
+  belongs_to :invitee, class_name: 'User'
+end
+
+# user1 = User.create(name: 'user1', email: 'user1@gmail.com', password: '123456');
+# user2 = User.create(name: 'user2', email: 'user2@gmail.com', password: '123456');
+# user3 = User.create(name: 'user3', email: 'user3@gmail.com', password: '123456');
+
+# user1 = User.first;
+# user2 = User.second;
+# user3 = User.last;
```
Hope you will remove this commented codes

descholar-ceo reviewed (changes requested):
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @bini-i  and @yocosaka  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your social media app is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] When I reject friend request, I should be able to request again after, and on the other side, the user from whom I requested friendship he/she should be able to send me another friendship request, currently your app is not behaving like that. Please fix it.
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

Commit:
Remove extra comment in friendship model.eb file

Commit:
Remove the condition in case of user rejected a request

yocosaka commented:
Hi, @descholar-ceo, Thanks for your great review, but I have one thing I'd like to ask you.
Should we fix it even if Your suggestion is not included in the requirements of this project?
At the beginning of this project, we make status boolean, and even if either user rejects a request, they was able to invite again. Now, we thought it as a little bit strange, that's why we intended that user who rejected or was rejected cannot send a request again.

yocosaka commented:
> Twitter

descholar-ceo commented:
Hello @yocosaka ,thank you for reaching out for more clarification. Yes actually you should make it to work, because that's a basic functionality of friendship request / rejection.

bini-i committed:
- Update migration for friendships, user model and users view _status
partial

- substitute status column with a boolean accepted for friendships table
- update user model to include accepted model rather than status column
- update status partial to allow rejected user to invite again

yocosaka commented:
@descholar-ceo, thanks a lot for your reply.
I understand. We tried to fix it following your suggestion.

descholar-ceo commented:
@yocosaka that's very commendable,

mahmoud717 reviewed (approved):
### STATUS: APPROVED ✔️ ✔️ 

**hi guys**

**Great job with making all of the requirements. the project is approved, you can go ahead and merge this pull request**
